mypy init

juanitorduz opened the pull request:
First steps towards adding mypy support, see  https://github.com/pymc-devs/pymc-bart/issues/76

juanitorduz commented on pymc_bart/tree.py:
```diff
@@ -52,66 +96,88 @@ class Tree:
+    def __init__(
+        self,
+        tree_structure: Dict[int, Node],
+        idx_leaf_nodes: Optional[List[int]],
+        output: Optional[npt.NDArray],
```
These two parameters could be `None` as we have a method returning `Tree(tree, None, None)`  below. So I added some additional changes to check The none condition before calling methods on these attributes (e.g. calling `append` and `remove`). This is one of the benefits of mypy :)

juanitorduz commented on pymc_bart/tree.py:
```diff
@@ -52,66 +96,88 @@ class Tree:
+    def __init__(
+        self,
+        tree_structure: Dict[int, Node],
+        idx_leaf_nodes: Optional[List[int]],
```
Do we want this to be a list or a numpy array? In the `pgbart` module we have something like 

```python
        self.a_tree = Tree.new_tree(
            leaf_node_value=init_mean / self.m,
            idx_data_points=np.arange(self.num_observations, dtype="int32"),
            num_observations=self.num_observations,
            shape=self.shape,
        )
```
which makes mypy fail when using `List[int]`

juanitorduz replied:
This generates the mypy error:
```python:
pymc_bart/pgbart.py:162: error: Argument "idx_data_points" to "new_tree" of "Tree" has incompatible type "ndarray[Any, dtype[Any]]"; expected "Optional[List[int]]"  [arg-type]
```

aloctavodia replied:
a numpy array

aloctavodia commented on pymc_bart/pgbart.py:
```diff
@@ -55,9 +110,9 @@ class PGBART(ArrayStepShared):
-        batch="auto",
-        model=None,
+        num_particles: int = 20,
+        batch: Union[str, int] = "auto",
```
Maybe settle on a tuple of floats with a default value of (0.1, 0.1)? This will require changes here https://github.com/pymc-devs/pymc-bart/blob/7c1c5bb4bd75cf2ca1545fb99e2d05d98cb1e0f8/pymc_bart/pgbart.py#L174-L181

juanitorduz replied:
done in https://github.com/pymc-devs/pymc-bart/pull/78/commits/b12feb528096f7fe097cbafe45f78669c4caddee

juanitorduz commented:
@aloctavodia I think I have made good progress. I still need to work on the docstrings and the rest of the files. Still, I think this is a good checkpoint to get some feedback as I am touching many places :) Don't feel pressured to review as I will continue working on it.

aloctavodia commented on pymc_bart/pgbart.py:
```diff
@@ -116,20 +168,15 @@ def __init__(
-                self.batch = batch
-            else:
-                self.batch = (batch, batch)
+        if batch == (0.1, 0.1):
+            batch_component = max(1, int(self.m * 0.1))
+            self.batch = (batch_component, batch_component)
```
This should be something like this. 


```suggestion
    batch_0 = max(1, int(self.m * batch[0]))
    batch_1 = max(1, int(self.m * batch[1]))
    self.batch = ( batch_0,  batch_1)
```

juanitorduz commented on pymc_bart/utils.py:
```diff
@@ -2,17 +2,29 @@
 
-def _sample_posterior(all_trees, X, rng, size=None, excluded=None):
+def _sample_posterior(
+    all_trees: List[List[Tree]],
```
I am not sure what the type of `all_trees` should be. In the `_sample_posterior` function we use it as ` shape = stacked_trees[0][0].predict(X[0]).size` so I tried with `List[List[Tree]]` however the `predict` method from the `Tree` class returns a `float` and that is why `mypy` complains about:
```python
pymc_bart/utils.py:58: error: "float" has no attribute "size" 
```

aloctavodia replied:
Predict should return an array with the number of elements determined by the first dimension of the shape argument passed to BART, so by default is only one element. But if we are fitting more than one BART variable like in the quartile regression example of the heteroscedasticity example this will have more elements

aloctavodia commented on pymc_bart/tree.py:
```diff
@@ -52,77 +106,102 @@ class Tree:
             for k, v in self.tree_structure.items()
         }
-        return Tree(tree, self.idx_leaf_nodes.copy(), self.output.copy())
+        idx_leaf_nodes = self.idx_leaf_nodes.copy() if self.idx_leaf_nodes is not None else None
+        output = self.output.copy() if self.output is not None else None
```
When this can be None?

aloctavodia replied:
thanks, I missed that. We "trim" trees before we send them to the Trace/InferenceData. We do this to reduce the memory footprint as we no longer use those trees for sampling, so we can change the Nones in `Tree(tree, None, None)` to anything we want. An empty list and array or a list with nan, or a sentinel value like -1...

juanitorduz replied:
ahhh perfect! This makes sense! This will make mypy (and me) happier.

juanitorduz replied:
fixed in https://github.com/pymc-devs/pymc-bart/pull/78/commits/8d4fc232cd415b0aa9384d6c5bd506fd2c94518e

aloctavodia commented on pymc_bart/tree.py:
```diff
@@ -52,77 +106,102 @@ class Tree:
-        return output.T
 
-    def predict(self, x, excluded=None):
+        if output is None:
```
again, isn't self.output always an array?

juanitorduz replied:
Not always, there is an edge case `Tree(tree, None, None)` https://github.com/pymc-devs/pymc-bart/blob/main/pymc_bart/tree.py#L102  If I am not missing something 🙈

juanitorduz replied:
(resolved) above.

juanitorduz replied:
see https://github.com/pymc-devs/pymc-bart/pull/78/commits/8d4fc232cd415b0aa9384d6c5bd506fd2c94518e

juanitorduz commented:
Test are passing locally but on the ci/cd we are getting
```python
 ModuleNotFoundError: No module named 'pymc.logprob.joint_logprob'
```
because of https://github.com/pymc-devs/pymc/pull/6599. (I will change the module name accordingly to match the latest release)

aloctavodia commented on pymc_bart/utils.py:
```diff
@@ -50,40 +64,46 @@ def _sample_posterior(all_trees, X, rng, size=None, excluded=None):
+    List[ax] : matplotlib axes
     """
-    ess_threshold = idata.posterior.chain.size * 100
+    ess_threshold = idata["posterior"]["chain"].size * 100
```
why this?

juanitorduz replied:
Unfortunately, `mypy` does not like `idata.posterior.chain.size` since `az.InferenceData` does not always have the `.posterior` sttribute
```python
❯ mypy
pymc_bart/utils.py:98: error: "InferenceData" has no attribute "posterior"  [attr-defined]
```

aloctavodia reviewed (commented):
LGTM!

juanitorduz commented:
> LGTM!

Great! I did change some small things which should not have any real impact on the code performance but feel free to test without rushing :)

aloctavodia commented on pymc_bart/pgbart.py:
```diff
@@ -422,7 +415,7 @@ def grow_tree(
     )
 
-    new_nodes = []
+    new_nodes = np.array([])
```
```suggestion
    new_nodes = np.empty(2, dtype=object)
```

aloctavodia commented on pymc_bart/tree.py:
```diff
@@ -52,164 +106,161 @@ class Tree:
+                )
             },
-            idx_leaf_nodes=[0],
+            idx_leaf_nodes=np.array([0]),
```
```suggestion
            idx_leaf_nodes=[0],
```

aloctavodia commented on pymc_bart/tree.py:
```diff
@@ -52,164 +106,161 @@ class Tree:
-        if node.is_leaf_node():
-            self.idx_leaf_nodes.append(index)
+        if node.is_leaf_node() and self.idx_leaf_nodes is not None:
+            # self.idx_leaf_nodes.append(index)
+            self.idx_leaf_nodes = np.append(self.idx_leaf_nodes, index)
```
```suggestion
            self.idx_leaf_nodes.append(index)
```

aloctavodia replied:
np.append is very slow compared to the append method from list, we should avoid it as we will need to perform this operation thousands of times

aloctavodia commented on pymc_bart/tree.py:
```diff
@@ -52,164 +106,161 @@ class Tree:
         current_node.idx_data_points = None
-        self.idx_leaf_nodes.remove(index_leaf_node)
+        if self.idx_leaf_nodes is not None:
+            # self.idx_leaf_nodes.remove(index_leaf_node)
+            self.idx_leaf_nodes = np.setdiff1d(self.idx_leaf_nodes, index_leaf_node)
```
```suggestion
            self.idx_leaf_nodes.remove(index_leaf_node)
```

juanitorduz replied:
added in https://github.com/pymc-devs/pymc-bart/pull/78/commits/b17d27e27292cf969890163075d30c9148674b33

aloctavodia commented on pymc_bart/tree.py:
```diff
@@ -52,164 +106,161 @@ class Tree:
         """
         if excluded is None:
-            excluded = []
+            excluded = np.array([])
```
can we keep the list? I don't think we need an array.

aloctavodia replied:
Not for operations like appending. But yes for typical "numerical" operations like computing trigonometric operations or vector operations. 

Sometimes also makes sense to write code in "pure python" and then use numba to speed-up the code. In many instances that can be 2x or faster than numpy/scipy (we have a few examples of that in PyMC-BART). Also when your function is very specialized like if all you are going to do is to perform operations with points in 3D space, an ad-hoc "dot" function can be faster than "np.dot".

juanitorduz replied:
Fixed in https://github.com/pymc-devs/pymc-bart/pull/78/commits/407e4a6d2619c7d71d4c250ab600db61b2b9e071

juanitorduz commented:
Thank you @aloctavodia ! I will look into your comments in detail!

juanitorduz commented:
@aloctavodia I addressed your comments... would you mind running the benchmark to see if we do not have performance degradation with these changes? Thanks :)

aloctavodia commented:
According to the benchmark, this may be 5% faster or 5% slower, haha.

Joking aside I think this is ready to merge, I am going to wait a little bit in case  @fjloyola has some comments to add.

juanitorduz commented:
Amazing! Thanks for the feedback!

juanitorduz commented:
Amazing! Thanks for the feedback!

aloctavodia commented:
Thanks @juanitorduz!!!
